wire-java-generator/src/main/java/com/squareup/wire/java/JavaGenerator.java

```diff
@@ -553,6 +592,61 @@ private TypeSpec generateEnclosingType(EnclosingType type) {
     return builder.build();
   }
 
+  /** Returns generated code that maps field names to field tags. */
+  public TypeSpec generateProtoFields(MessageType type) {
+    NameAllocator nameAllocator = nameAllocators.getUnchecked(type);
+
+    ClassName javaType = (ClassName) protoFieldsTypeName(type.type());
+    TypeSpec.Builder builder = TypeSpec.classBuilder(javaType.simpleName() + "ProtoFields");
+    builder.addModifiers(PUBLIC, FINAL);
+
+    for (Type nestedType : type.nestedTypes()) {
+      if (!protoTypeToAdapter.containsKey(nestedType.type())) {
+        throw new IllegalArgumentException("Missing custom proto adapter for "
+            + nestedType.type().enclosingTypeOrPackage() + "." + nestedType.type().simpleName()
+            + " when enclosing proto has custom proto adapter.");
+      }
+      if (nestedType instanceof MessageType) {
+        builder.addType(generateProtoFields((MessageType) nestedType));
+      }
+    }
+
+    for (Field field : type.fieldsAndOneOfFields()) {
+      String fieldName = nameAllocator.get(field);
+
+      TypeName typeName =
+          ParameterizedTypeName.get(ClassName.get(ProtoField.class), fieldType(field));
+      FieldSpec.Builder fieldBuilder =
+          FieldSpec.builder(typeName, fieldName, PUBLIC, STATIC, FINAL);
+
+      CodeBlock adapter;
+      if (field.type().isMap()) {
```

Comment:
can you make the existing `adapterFor()` method smarter and handle this? (I’m unsure)